river/neighbors/knn_classifier.py

```diff
@@ -1,176 +1,193 @@
-    numerical.
+    See the NearestNeighbors documentation for details about the base model,
+    along with KNNBase for an example of providing your own distance function.
+    Note that since the window is moving and we keep track of all classes that
+    are added at some point, a class might be returned in a result (with a
+    value of 0) if it is no longer in the window. You can call
+    model.clean_up_classes(), or set `cleanup_every` to a non-zero value.
 
     Examples
     --------
-    >>> from river import datasets
-    >>> from river import evaluate
-    >>> from river import metrics
-    >>> from river import neighbors
-    >>> from river import preprocessing
-
+    >>> from river import datasets, neighbors, preprocessing
+    >>> from river import evaluate, metrics
     >>> dataset = datasets.Phishing()
 
     >>> model = (
     ...     preprocessing.StandardScaler() |
     ...     neighbors.KNNClassifier()
     ... )
 
-    >>> metric = metrics.Accuracy()
-
-    >>> evaluate.progressive_val_score(dataset, model, metric)
-    Accuracy: 88.07%
+    >>> for x, y in dataset.take(100):
+    ...     model = model.learn_one(x, y)
 
+    >>> for x, y in dataset.take(1):
+    ...     model.predict_one(x)
+    {False: 0.0, True: 1.0}
     """
 
     def __init__(
         self,
         n_neighbors: int = 5,
         window_size: int = 1000,
-        leaf_size: int = 30,
-        p: float = 2,
+        min_distance_keep: float = 0.0,
         weighted: bool = True,
-        **kwargs
+        cleanup_every: int = 0,
+        distance_func: DistanceFunc = None,
+        softmax: bool = False,
     ):
         super().__init__(
             n_neighbors=n_neighbors,
             window_size=window_size,
-            leaf_size=leaf_size,
-            p=p,
-            **kwargs
+            min_distance_keep=min_distance_keep,
+            distance_func=distance_func,
         )
         self.weighted = weighted
-        self.classes_: typing.Set = set()
-        self.kwargs = kwargs
+        self.cleanup_every = cleanup_every
+        self.classes = set()
+        self.softmax = softmax
+        self._cleanup_counter = cleanup_every
 
-    def _unit_test_skips(self):
-        return {"check_emerging_features", "check_disappearing_features"}
+    @property
+    def _multiclass(self):
+        return True
 
-    def learn_one(self, x, y):
-        """Update the model with a set of features `x` and a label `y`.
+    def clean_up_classes(self) -> "KNNClassifier":
+        """
+        Classes that are added (and removed) from the window may no longer be valid.
+        This method cleans up the window and and ensures only known classes
+        are added, and we do not consider "None" a class. It is called every
+        `cleanup_every` step, or can be called manually.
+        """
+        self.classes = {x for x in self.window if x[0][1] is not None}
+
+    def learn_one(self, x: dict, y=None):
```

Comment:
You don't have to type hint `x`, it's inherited. As for y, it shouldn't default to `None` because this is a supervised model.